README.md

```diff
@@ -198,289 +189,492 @@ e.increment();  // logs 1
+The `init` keyword for methods can also be used to build a `@bound` decorator, used as follows:
 
 ```js
 class C {
-  @initialize(f) a = b;
+  #x = 1;
+  @bound init method() { return this.#x; }
 }
+
+let c = new C;
+let m = c.method;
+m();  // 1, not TypeError
 ```
 
-is roughly equivalent to the following:
+The `@bound` decorator can be implemented as follows:
 
 ```js
-class C {
-  constructor() {
-    f.call(this, f, b, "a");
-  }
+function bound(method, {kind, name}) {
+  assert(kind === "init-method");
+  return {method, initialize() { this[name] = this[name].bind(this); }};
 }
 ```
 
-Details:
-- `@initialize` can only be used on public field declarations (including symbols/computed property names), not private.
-- `@initialize` can be used with both static and instance fields.
-- The return value of the callback must be undefined.
+## Possible extensions
+
+Decorators on further constructs are investigated in [EXTENSIONS.md](./EXTENSIONS.md).
 
-## User-defined Decorators
+# Syntax
 
-JavaScript programmers can make their own decorators by composing built-in decorators.
+This decorators proposal uses the syntax of the previous Stage 2 decorators proposal. This means that:
+- Decorator expressions are restricted to a chain of variables, property access with `.` but not `[]`, and calls `()`. To use an arbitrary expression as a decorator, `@(expression)` is an escape hatch.
+- Class expressions may be decorated, not just class declarations.
+- Class decorators come after `export` and `default`.
 
-### `decorator @xyz` declarations
+There is no special syntax for defining decorators; any function can be applied as a decorator.
 
-Decorators may be defined as a simple composition of other decorators. You can use all the fancy JavaScript features you want inside the arguments, but at the top level, this is just a string of decorators and arguments for these decorators. There's no way to conditionally use one decorator in one situation and another in another situation, for example.
+# Detailed semantics 
 
-Example:
+The three steps of decorator evaluation:
+1. Decorator expressions (the thing after the `@`) are *evaluated* interspersed with computed property names.
+1. Decorators are *called* (as functions) during class definition, after the methods have been evaluated but before the constructor and prototype have been put together.
+1. Decorators are *applied* (mutating the constructor and prototype) all at once, after all of them have been called.
+
+The semantics here generally follow the consensus at the May 2016 TC39 meeting in Munich.
+
+## 1. Evaluating decorators
+
+Decorators are evaluated as expressions, interspersed in their evaluation order with computed property names. This goes left to right, top to bottom. The result of decorators is stored in the equivalent of local variables to be later called after the class definition initially finishes executing.
```

Comment:
Is `interspersed` a typo of `interpreted`?

Reply:
After re-reading the beginning of "Detailed semantics" I can see how ["interspersed" means "placed among"](https://www.merriam-webster.com/dictionary/intersperse) in this case.

As a non-native speaker, the unusual word is a bit confusing. It might be fine to keep as it is, if it's well known a jargon in proposals.

Reply:
I'd like to ensure this proposal is accessible by lots of different audiences, including non-native English speakers. Now that you understand what this sentence is getting at, could you suggest another wording that would be more clear?

Reply:
The following sentence is the way I understand its original meaning. Feel free to improve on it:

> Decorators are evaluated as expressions, being ordered along with computed property names.